compat/mingw.c

```diff
@@ -426,6 +426,43 @@ static wchar_t *normalize_ntpath(wchar_t *wbuf)
 	return wbuf;
 }
 
+/* rename and unlink a file. This is needed for the ability to recreate a file
+ * with the same name, for example on checkout which unlinks and recreates the file */
+static int rename_and_unlink(const wchar_t *pathname, int (*unlinker)(const wchar_t *))
+{
+	wchar_t temppath[MAX_LONG_PATH];
+	*temppath = 0;
+	/* leave some space for the .unlink_XXXXXX extension */
+	wcsncat(temppath, pathname, sizeof(temppath) - 15);
+	wcsncat(temppath, L".unlink_XXXXXX", sizeof(temppath));
+	_wmktemp(temppath);
+	if (MoveFileExW(pathname, temppath, 0)) {
+		const int res = unlinker(temppath);
+		if (res) {
+			const DWORD err = GetLastError();
+			MoveFileExW(temppath, pathname, 0);
+			SetLastError(err);
+		}
+		return res;
+	}
+	return unlinker(pathname);
+}
+
+static int do_unlink(const wchar_t *wpathname, int (*unlinker)(const wchar_t *))
```

Comment:
If we do not extract the rename from this function, then the function name is not appropriate any more. Maybe `remove_safely`?

Reply:
Named safe_unlink